lib/ansible/modules/cloud/amazon/iot_thing.py

```diff
@@ -0,0 +1,484 @@
+      setting attributes and thing types.  It also allows principals to be attached and detached.
+version_added: "2.5"
+requirements: [ boto3 ]
+author:
+    - 'Andrew Kowpak (@akowpak-miovision)'
+options:
+  name:
+    description:
+      - The name of the IOT thing.
+    required: true
+  state:
+    description:
+      - If the IOT thing should be added or removed.
+    default: present
+    choices: [ 'present', 'absent' ]
+  type:
+    description:
+      - An IOT thing type to associate with this IOT thing.
+      - This parameter is mutually exclusive with I(remove_type).
+  remove_type:
+    description:
+      - If the IOT type association should be removed from this IOT thing.
+      - This parameter is mutually exclusive with I(type).
+    type: bool
+  attached_principals:
+    description:
+      - List of certificate or credential ARNs that should be attached to this IOT thing.
+  detached_principals:
+    description:
+      - List of certificate or credential ARNs that should not be attached to this IOT thing.
+  attributes:
+    description:
+      - A dictionary of attributes that should be set on this IOT thing.
+  absent_attributes:
+    description:
+      - A list of attribute names that should not exist on this IOT thing.
+  expected_version:
+    description:
+      - The version of the IOT thing that is expected to exist.  If a different version exists, then an attempt to
+        update or delete the IOT thing will fail.
+extends_documentation_fragment:
+    - aws
+    - ec2
+'''
+
+
+EXAMPLES = '''
+- name: create IOT thing
+  iot_thing:
+    name: mything
+    state: present
+
+- name: delete IOT thing
+  iot_thing:
+    name: mything
+    state: absent
+
+- name: set type of IOT thing
+  iot_thing:
+    name: mything
+    type: mytype
+
+- name: set the type associated with an IOT thing
+  iot_thing:
+    name: mything
+    remove_type: True
+
+- name: disassociate an IOT thing from a type
+  iot_thing:
+    name: mything
+    type:
+      state: absent
+
+- name: attach a certificate to an IOT thing
+  iot_thing:
+    name: mything
+    attached_principals:
+      - "arn:aws:iot:us-east-1:account-id:cert/uuid"
+
+- name: detach a certificate from an IOT thing
+  iot_thing:
+    name: mything
+    detached_principals:
+      - "arn:aws:iot:us-east-1:account-id:cert/uuid"
+
+- name: add attributes to an IOT thing
+  iot_thing:
+    name: mything
+    attributes:
+      key1: value1
+      key2: value2
+
+- name: remove attributes from an IOT thing
+  iot_thing:
+    name: mything
+    absent_attributes:
+      - key3
+'''
+
+
+RETURN = '''
+default_client_id:
+    description: The ID of the default client for this IOT thing.
+    returned: on success when I(state=present)
+    type: string
+    sample: mythingid
+version:
+    description: The current version of the thing record in the IOT registry.
+    returned: on success when I(state=present)
+    type: int
+    sample: 123
+attribute:
```

Comment:
Pretty sure this should be attributes? http://boto3.readthedocs.io/en/latest/reference/services/iot.html#IoT.Client.describe_thing